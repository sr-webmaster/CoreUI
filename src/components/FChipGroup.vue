<template>
  <v-card class="f-chip-group">
    <v-card-title v-if="hasTitle">
      {{ title }}
    </v-card-title>
    <v-divider v-if="hasTitle" />
    <v-card-text>
      <f-chip
        v-for="item in tagsComputed"
        :key="item.id"
        :color="isSelected(item.id) ? 'primary' : 'accent'"
        @click.prevent="changeTags(item)"
      >
        {{ item.name }}
      </f-chip>
    </v-card-text>
  </v-card>
</template>

<script>
import remove from 'lodash/remove'
import FChip from './FChip'

export default {
  components: {
    FChip
  },
  props: {
    title: {
      type: String,
      default: ''
    },
    tags: {
      type: Array,
      required: true
    },
    selected: {
      type: Array,
      default: () => []
    }
  },
  data () {
    return {
      selectedTagsData: [...this.selected]
    }
  },
  computed: {
    hasTitle () {
      return this.title.length > 0
    },
    tagsComputed () {
      return this.tags
    }
  },
  watch: {
    selected (val) {
      this.selectedTagsData = val
    }
  },
  methods: {
    changeTags (tag) {
      if (!this.isSelected(tag.id)) {
        this.selectedTagsData.push(tag)
      } else {
        this.selectedTagsData = remove(this.selectedTagsData, function (item) {
          return !(item.id === tag.id)
        })
      }
      this.$emit('input', this.selectedTagsData)
    },
    isSelected (id) {
      return this.selectedTagsData.some(element => element.id === id)
    }
  }
}
</script>

<style lang="scss">
@import '@freshinup/core-ui/src/styles/sass/_variables';
.title {
  font-size: $fresh-typography--title-link !important;
}
.title:link {
  text-decoration: none;
}
</style>
