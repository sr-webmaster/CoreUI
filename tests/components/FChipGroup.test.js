import { mount } from '@vue/test-utils'
import Component from '@freshinup/core-ui/src/components/FChipGroup.vue'
import { FIXTURE_UI_FCHIPGROUP_TAGS } from '@freshinup/core-ui/tests/__data__/FChipGroup/tags'

describe('FChipGroup', () => {
  describe('Visuals', () => {
    it('snapshot default', async () => {
      const wrapper = mount(Component, {
        propsData: {
          selected: [],
          tags: FIXTURE_UI_FCHIPGROUP_TAGS
        }
      })
      expect(wrapper.element).toMatchSnapshot()
    })
    it('snapshot with title', async () => {
      const wrapper = mount(Component, {
        propsData: {
          selected: [],
          tags: FIXTURE_UI_FCHIPGROUP_TAGS,
          title: 'This is with a title',
          titleClass: 'grey--text text-uppercase font-weight-bold'
        }
      })
      expect(wrapper.element).toMatchSnapshot()
    })
  })
  describe('Property Change', () => {
    it('selected', async () => {
      const initialTags = [
        FIXTURE_UI_FCHIPGROUP_TAGS[0],
        FIXTURE_UI_FCHIPGROUP_TAGS[1],
        FIXTURE_UI_FCHIPGROUP_TAGS[2]
      ]

      const changedTags = [
        FIXTURE_UI_FCHIPGROUP_TAGS[0],
        FIXTURE_UI_FCHIPGROUP_TAGS[1]
      ]

      const wrapper = mount(Component, {
        propsData: {
          selected: initialTags,
          tags: FIXTURE_UI_FCHIPGROUP_TAGS
        }
      })

      expect(wrapper.vm.selectedTagsData).toEqual(initialTags)

      wrapper.setProps({ selected: changedTags })

      await wrapper.vm.$nextTick()
      await wrapper.vm.$nextTick()

      expect(wrapper.vm.selectedTagsData).toEqual(changedTags)
    })
  })
  describe('Methods', () => {
    describe('changeTags()', () => {
      test('adds selected and emits "input" with selectedTags', () => {
        const wrapper = mount(Component, {
          propsData: {
            selected: [],
            tags: FIXTURE_UI_FCHIPGROUP_TAGS
          }
        })
        wrapper.vm.changeTags(FIXTURE_UI_FCHIPGROUP_TAGS[0])
        expect(wrapper.vm.selectedTagsData).toEqual([FIXTURE_UI_FCHIPGROUP_TAGS[0]])
        // assert event has been emitted
        expect(wrapper.emitted('input')).toBeTruthy()
        expect(wrapper.emitted('input')).toHaveLength(1)
        expect(wrapper.emitted('input')[0]).toEqual([[FIXTURE_UI_FCHIPGROUP_TAGS[0]]])
        wrapper.vm.changeTags(FIXTURE_UI_FCHIPGROUP_TAGS[1])
        expect(wrapper.emitted('input')).toHaveLength(2)
        expect(wrapper.emitted('input')[1]).toEqual([[FIXTURE_UI_FCHIPGROUP_TAGS[0], FIXTURE_UI_FCHIPGROUP_TAGS[1]]])
      })
      test('removes previously selected and emits "input" with selectedTags', () => {
        const wrapper = mount(Component, {
          propsData: {
            selected: [FIXTURE_UI_FCHIPGROUP_TAGS[0]],
            tags: FIXTURE_UI_FCHIPGROUP_TAGS
          }
        })
        wrapper.vm.changeTags(FIXTURE_UI_FCHIPGROUP_TAGS[0])
        expect(wrapper.vm.selectedTagsData).toEqual([])
        // assert event has been emitted
        expect(wrapper.emitted('input')).toBeTruthy()
        expect(wrapper.emitted('input')).toHaveLength(1)
        expect(wrapper.emitted('input')[0]).toEqual([[]])
      })
      describe('isSelected()', () => {
        test('returns true', () => {
          const wrapper = mount(Component, {
            propsData: {
              selected: [FIXTURE_UI_FCHIPGROUP_TAGS[0]],
              tags: FIXTURE_UI_FCHIPGROUP_TAGS
            }
          })
          expect(wrapper.vm.isSelected(FIXTURE_UI_FCHIPGROUP_TAGS[0].id)).toEqual(true)
        })
      })
    })
  })
})
