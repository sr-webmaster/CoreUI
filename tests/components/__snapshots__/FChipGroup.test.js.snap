// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FChipGroup Visuals snapshot default 1`] = `
<div
  class="f-chip-group v-card v-sheet theme--light"
>
  <!---->
   
  <!---->
   
  <div
    class="v-card__text"
  >
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 1
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 2
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 3
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 4
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 5
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 6
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 7
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 8
    
      </span>
    </span>
  </div>
</div>
`;

exports[`FChipGroup Visuals snapshot with title 1`] = `
<div
  class="f-chip-group v-card v-sheet theme--light"
  titleclass="grey--text text-uppercase font-weight-bold"
>
  <div
    class="v-card__title"
  >
    
    This is with a title
  
  </div>
   
  <hr
    class="v-divider theme--light"
  />
   
  <div
    class="v-card__text"
  >
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 1
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 2
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 3
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 4
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 5
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 6
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 7
    
      </span>
    </span>
    <span
      class="v-chip theme--light accent accent-text--text"
      tabindex="0"
    >
      <span
        class="v-chip__content"
      >
        
      Category 8
    
      </span>
    </span>
  </div>
</div>
`;
